File: Playwright/examples/nodejs/cross_match_reveal.mjs
/**
 * Reference: ContactOut cross-match + 429 backoff (Node.js / Playwright).
 * NOT wired to production — see Playwright/README.md
 *
 * Usage:
 *   npm init -y && npm i playwright
 *   node examples/nodejs/cross_match_reveal.mjs "Ryan Cronin" "https://linkedin.com/in/..."
 */

import { chromium } from "playwright";

const humanWait = (min = 3000, max = 7000) =>
  new Promise((r) => setTimeout(r, Math.floor(Math.random() * (max - min + 1)) + min));

function cleanLinkedIn(url) {
  return (url || "").split("?")[0].toLowerCase().replace(/\/$/, "");
}

async function gotoWithRetry(page, url, maxAttempts = 5) {
  let backoff = 5000;
  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    await humanWait();
    const response = await page.goto(url, { waitUntil: "domcontentloaded" });
    if (response?.status() === 429) {
      console.warn(`[429] Backing off ${backoff}ms`);
      await new Promise((r) => setTimeout(r, backoff));
      backoff *= 2;
      continue;
    }
    return response;
  }
  throw new Error(`429 loop on ${url}`);
}

async function verifyCredits(page) {
  const body = await page.innerText("body");
  const match = body.match(/(\d+)\s*credits?\s*(left|remaining)/i);
  if (match) return parseInt(match[1], 10);
  return 999;
}

async function crossMatchAndReveal(targetName, expectedLinkedIn) {
  const browser = await chromium.launch({
    headless: false,
    args: ["--disable-blink-features=AutomationControlled"],
    // proxy: { server: process.env.CONTACTOUT_PROXY_URL },
  });

  const context = await browser.newContext({
    storageState: process.env.CONTACTOUT_SESSION_FILE,
    userAgent:
      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
    viewport: { width: 1440, height: 900 },
  });
  const page = await context.newPage();

  try {
    await gotoWithRetry(page, "https://contactout.com/dashboard/search");
    const credits = await verifyCredits(page);
    if (credits <= 0) {
      console.error("Credits depleted — stopping");
      return;
    }

    const nameInput = page.getByLabel("Name").first();
    await nameInput.click();
    await nameInput.fill(targetName);
    await nameInput.press("Enter");
    await humanWait(3000, 6000);
    await page.mouse.wheel(0, 400);

    const cards = await page.locator("table tbody tr, .profile-card").all();
    const expected = cleanLinkedIn(expectedLinkedIn);

    for (const card of cards) {
      const anchor = card.locator('a[href*="linkedin.com"]').first();
      if ((await anchor.count()) === 0) continue;
      const found = cleanLinkedIn(await anchor.getAttribute("href"));
      if (found !== expected) continue;

      console.log("LinkedIn verified — scoped reveal");
      const reveal = card.getByRole("button", { name: /reveal|show email/i });
      await humanWait(1500, 3500);
      await reveal.first().click();
      break;
    }
  } finally {
    await context.close();
    await browser.close();
  }
}

const [, , name, linkedin] = process.argv;
if (!name || !linkedin) {
  console.error("Usage: node cross_match_reveal.mjs <name> <linkedin_url>");
  process.exit(1);
}
crossMatchAndReveal(name, linkedin);

File: Playwright/examples/python/cross_match_reveal.py
#!/usr/bin/env python3
"""
Reference: async ContactOut cross-match + 429 backoff (Python / Playwright).
NOT wired to production — prefer scripts/test_cross_match.py for this package.

Usage:
  pip install playwright && playwright install chrome
  python examples/python/cross_match_reveal.py "Ryan Cronin" "https://linkedin.com/in/..."
"""
from __future__ import annotations

import asyncio
import random
import re
import sys
from pathlib import Path

from playwright.async_api import async_playwright


async def human_wait(min_ms: int = 3000, max_ms: int = 7000) -> None:
    await asyncio.sleep(random.uniform(min_ms / 1000, max_ms / 1000))


def clean_linkedin(url: str | None) -> str:
    return (url or "").split("?")[0].lower().rstrip("/")


async def goto_with_retry(page, url: str, max_attempts: int = 5):
    backoff = 5.0
    for attempt in range(max_attempts):
        await human_wait()
        response = await page.goto(url, wait_until="domcontentloaded")
        if response and response.status == 429:
            print(f"[429] Vanishing for {backoff}s (attempt {attempt + 1})")
            await asyncio.sleep(backoff)
            backoff = min(backoff * 2, 900)
            continue
        return response
    raise RuntimeError(f"429 loop on {url}")


async def verify_credits(page) -> int:
    body = await page.inner_text("body")
    match = re.search(r"(\d+)\s*credits?\s*(left|remaining)", body, re.I)
    return int(match.group(1)) if match else 999


async def cross_match_and_reveal(name: str, expected_linkedin: str) -> None:
    session = Path(__file__).resolve().parents[2] / ".contactout-session.json"
    storage = str(session) if session.is_file() else None

    async with async_playwright() as p:
        browser = await p.chromium.launch(
            headless=False,
            args=["--disable-blink-features=AutomationControlled"],
        )
        context = await browser.new_context(
            storage_state=storage,
            viewport={"width": 1440, "height": 900},
        )
        page = await context.new_page()
        try:
            await goto_with_retry(page, "https://contactout.com/dashboard/search")
            if await verify_credits(page) <= 0:
                print("Credits depleted")
                return

            name_input = page.get_by_label("Name").first()
            await name_input.click()
            await name_input.fill(name)
            await name_input.press("Enter")
            await human_wait(3000, 6000)
            await page.evaluate("window.scrollBy(0, 350)")

            expected = clean_linkedin(expected_linkedin)
            cards = await page.locator("table tbody tr, .profile-card").all()
            for card in cards:
                anchor = card.locator('a[href*="linkedin.com"]').first()
                if await anchor.count() == 0:
                    continue
                found = clean_linkedin(await anchor.get_attribute("href"))
                if found != expected:
                    continue
                print("Match verified — reveal on card")
                reveal = card.get_by_role("button", name=re.compile(r"reveal|show email", re.I))
                await human_wait(1500, 3500)
                await reveal.first.click()
                await human_wait()
                break
        finally:
            await context.close()
            await browser.close()


if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Usage: python cross_match_reveal.py <name> <linkedin_url>")
        raise SystemExit(1)
    asyncio.run(cross_match_and_reveal(sys.argv[1], sys.argv[2]))
